### PR TITLE
Async Read for PG and ObjectStores

### PR DESCRIPTION
**Pull request by v-sinha**
I had a prior PR (https://github.com/ceph/ceph/pull/6776) on the same topic but too many changes since then so that is marked closed now.

The purpose of this is to enable an asynchronous, callback-based interface and functionality for sending Reads to Objectstore, and then to implement an Asynchronous (Linux AIO) mechanism within FileStore that can be used to serve the incoming Async requests from the PG layer. This type of functionality is probably most beneficial to fast storage such as SSDs.

A brief description of the implementation is as follows.

- Existing Sync interfaces are unchanged. ObjectStores that do not implement Async Read functionality are not affected.

- In do_osd_ops() and related routines, a callback mechanism is enabled by passing in a callback argument. Then if the PGBackend reports that Async is available, Async processing will occur and invoke the callback as expected. Otherwise do_osd_ops() executes the callback itself (in Sync mode). The interface to the caller is the same in either case. Further, for small IO small IO (currently set to less than 256 bytes) the Async mechanism is not invoked. If there is only one callback expected, then additional locking overhead is avoided.

- The existing ObjectStore interface for doing synchronous Reads is unchanged. Async operation is disabled by default in ceph.conf. For each instance of ObjectStore, Async may be enabled/disabled individually.

- In FileStore, Async completion threads are created on mount(), if functionality is enabled. If filestore_async_threads=0 then the FileStore will not operate in Async mode (or create any Async completion threads). The Async completion threads just wait for AIO completions using io_getevents() and complete processing directly themselves (i.e. without going back to the sharding thread).

- I found that unless Direct IO (O_DIRECT) is used the performance gains due to Async functionality are not realized. This is consistent with what one finds about the Linux kernel implementation of AIO. So this code enables the use of O_DIRECT for Async Reads, by maintaining a separate FDCache ("fdcache_direct") in additional to the usual FileStore::fdcache, in which file descriptors are created by opening files with O_DIRECT. Again, this is avoided if Async functionality is not enabled.

Comments and feedback appreciated.
Thanks,
Vikas

### CODE REVIEW

**Comment by athanatos** [comment]
You seem to have 4 extra commits here.  Can you rebase cleanly on master?

File: src/os/ObjectStore.h [hunk 1]
```diff
@@ -2011,7 +2011,18 @@ class ObjectStore {
...
+   virtual int async_read_dispatch(
+     Context *ctx,
+     const coll_t& cid,
+     const ghobject_t& oid,
+     uint64_t offset,
+     size_t len,
+     bufferlist* bl,
+     uint32_t op_flags = 0,
+     bool allow_eio = false,
+     bool directio = false) {return -EWOULDBLOCK;}
```
Comment (athanatos) [inline src/os/ObjectStore.h]: spaces around the braces. (and below)

File: src/os/filestore/FileStore.cc [hunk 2]
```diff
@@ -231,16 +231,24 @@ int FileStore::lfn_open(const coll_t& cid,
...
 			FDRef *outfd,
-                        Index *index)
+			Index *index,
+			bool directio)
 {
   assert(outfd);
   int r = 0;
   bool need_lock = true;
   int flags = O_RDWR;
+  FDCache *cachep;
```
Comment (athanatos) [inline src/os/filestore/FileStore.cc]: At least  move this adjacent to the block which actually sets it.  I'd rather you actually did

```c++
FDCache *cachep = directio ? &fdcache_direct : &fdcache;
````

That would reduce the likelihood of someone later using it before it's initialized.

**Comment by athanatos** [comment]
There is no need to make object classes capable of doing async reads just to make get_id to an async read.  get_id is used once on image open -- not during the common read path.

**Comment by v-sinha** [comment]
Originally I saw a lot of Reads via get_id, and it seemed to me that it was one per object (not per RBD image). It was small-sized as well (less than 32 bytes).
In this version I have added code to avoid Async for small Reads, so this is not getting executed (just verified it). So I will remove it.

**Comment by v-sinha** [comment]
I plan to close this one and open a new one that is based on master.

**Pull request closed without merging**